### PR TITLE
QUIC: Add HTTP/3 demo using nghttp3

### PR DESCRIPTION
**Pull request by hlandau**
Fixes https://github.com/openssl/project/issues/41

### CODE REVIEW

File: demos/http3/ossl-nghttp3.c [hunk 1]
```diff
@@ -0,0 +1,603 @@
...
+        if (SSL_set_alpn_protos(conn->qconn, alpn, sizeof(alpn))) {
+            /* SSL_set_alpn_protos returns 1 on failure */
+            ERR_raise_data(ERR_LIB_USER, ERR_R_INTERNAL_ERROR,
+                           "failed to configure ALPN");
+            goto err;
+        }
+
+    BIO_set_nbio(conn->qconn_bio, 1);
+
+    if (!SSL_set_default_stream_mode(conn->qconn, SSL_DEFAULT_STREAM_MODE_NONE)) {
```
Comment (mattcaswell) [inline demos/http3/ossl-nghttp3.c:261]: Some extra comments around this line and the `BIO_set_nbio` call above to explain what is going on might be useful

File: demos/http3/ossl-nghttp3.c [hunk 2]
```diff
@@ -0,0 +1,603 @@
...
+        goto err;
+
+    bare_hostname = BIO_get_conn_hostname(qconn_bio);
+    if (bare_hostname == NULL)
+        goto err;
+
+    if (BIO_get_ssl(qconn_bio, &qconn) == 0)
+        goto err;
+
+    if (SSL_set1_host(qconn, bare_hostname) <= 0)
```
Comment (mattcaswell) [inline demos/http3/ossl-nghttp3.c:359]: Shouldn't we also call `SSL_set_tlsext_host_name()` here?

**Comment by t8m** [comment]
Wow @hlandau, great work!

File: demos/http3/ossl-nghttp3.c (hunk 2 above)
Comment (hlandau) [reply demos/http3/ossl-nghttp3.c:359]: We do this in ddd-03 (e.g. direct FD) where it's necessary but it's not necessary when using the ssl_connect BIO, right?

TLS 1.3 requires SNI so it wouldn't even work without it, no...?

**Comment by hlandau** [comment]
Added some more comments.

**Comment by t-j-h** [comment]
All the usual project copyright headers need to go around these files too ... you can be sure this code will end up copied into various applications in the short term so are those prefixes reasonable? H3_ seems unwise - perhaps OSSL_DEMO_H3? We don't want to unnecessarily make the forward path harder here.

A direct link to https://github.com/ngtcp2/nghttp3 should be included.

**Comment by hlandau** [comment]
@t-j-h Updated

**Comment by hlandau** [comment]
Updated. The lsquic issue is fixed, turns out we weren't sending SNI. Their error returned ("internal error") is suboptimal but so be it.

**Review by mattcaswell — APPROVED** [review]

**Review by t-j-h — APPROVED** [review]

**Pull request closed without merging**